### ISSUE TITLE
spark.task.cpus > 1 causes not all index partitions are persisted (data loss)

### ISSUE THREAD

**Issue by perrital** [jelmerk/hnswlib#63]
brute_force = BruteForceSimilarity(
    identifierCol='id',
    queryIdentifierCol='id',
    featuresCol='embedding',
    distanceFunction='cosine',
    k=200,
    numReplicas=0,
    numPartitions=15,
    excludeSelf=False,
    predictionCol='prediction'
)

model = brute_force.fit(q_embeddings)
model.write().overwrite().save(inference_base_path)

playing with spark.task.cpus and parallelism causes random appearing loss of index partitions 
working on k8s cluster

**Comment by perrital** [comment]
Investigating more, it seems like the indexes are generated on the executors, but don't find their way to s3

**Comment by jelmerk** [comment]
What files are being created under q_embeddings ?

**Comment by perrital** [comment]
q_embeddings.printSchema()
root
 |-- id: integer (nullable = true)
 |-- embedding: vector (nullable = true)


q_embeddings.take(1)
[Row(id=144808529, embedding=DenseVector([-0.0717, 0.0678, -0.0391, -0.0838, 0.012, 0.0936, -0.204, -0.1352, 0.0109, -0.1168, 0.1239, -0.0726, 0.2306, -0.0213, 0.1084, 0.2915, -0.0495, -0.1308, -0.09, -0.0206, 0.1202, 0.0322, -0.1932, 0.0782, -0.2072, -0.1579, -0.1023, 0.0924, -0.1313, -0.021, 0.134, -0.0277, 0.0361, -0.1267, -0.1022, 0.0558, -0.169, -0.1385, 0.058, -0.2013, 0.0136, -0.1415, -0.1482, -0.0933, -0.1589, 0.0956, 0.2642, -0.1398, 0.0138, -0.1669, 0.0818, 0.1304, -0.1037, -0.0792, -0.0056, 0.0734, -0.0038, -0.2278, -0.1226, -0.0627, -0.1865, 0.1155, -0.0469, 0.1591]))]

**Comment by jelmerk** [comment]
can you list the files in the folder inside the s3 bucket you write to

**Comment by perrital** [comment]
executor_cores: 2,
    executor_instances: 4,
    task_cpus: 2,
    executor_memory: "20G",
    enable_dynamic_resource_allocation: "false",
    spark.speculation': 'false',
    'spark.task.maxFailures': '1',
    "spark.hnswlib.settings.index.cache_folder": '/tmp/tal',
    'spark.scheduler.minRegisteredResourcesRatio': '1',
    'spark.scheduler.maxRegisteredResourcesWaitingTime': f'{5 * 60 * 1000}',

2023-06-12 15:25:47     532272 devs/gnn/test/inference/metal/ranking/tfmodel/indices/1
2023-06-12 15:25:47     488980 devs/gnn/test/inference/metal/ranking/tfmodel/indices/2
2023-06-12 15:25:46          0 devs/gnn/test/inference/metal/ranking/tfmodel/metadata/_SUCCESS
2023-06-12 15:25:46        448 devs/gnn/test/inference/metal/ranking/tfmodel/metadata/part-00000

**Comment by jelmerk** [comment]
can you cat the contents of `devs/gnn/test/inference/metal/ranking/tfmodel/metadata/part-00000`

and how many elements did you put in the index ?

**Comment by perrital** [comment]
{"class":"com.github.jelmerk.spark.knn.bruteforce.BruteForceSimilarityModelImpl","timestamp":1686572743855,"sparkVersion":"3.3.2","uid":"BruteForceSimilarity_23c56c637cfc","identifierType":"int","vectorType":"vector","partitions":4,"paramMap":{"numReplicas":0,"queryIdentifierCol":"id","queryPartitionsCol":"part","featuresCol":"embedding","similarityThreshold":-1.0,"outputFormat":"full","predictionCol":"prediction","excludeSelf":false,"k":200}}


3760 elements

**Comment by perrital** [comment]
I see in spark UI that the data is sent to the executors - but only on part of them the index get actually built 
also used query partition to navigate the search to specific partition - I see that spark send the record to the correct node, but randomly do the query get executed on it

**Comment by jelmerk** [comment]
hmm whats in that manifest is different than the config you shared before though, partitions is 4 instead of 15,  and i see a queryPartitionsCol in there so you are manually assigning the partitions in that

**Comment by perrital** [comment]
sorry.... I've run it so many times :) 
did a clean run again with 4 partitions you can see that only one was persisted 

{"class":"com.github.jelmerk.spark.knn.bruteforce.BruteForceSimilarityModelImpl","timestamp":1686576688130,"sparkVersion":"3.3.2","uid":"BruteForceSimilarity_7388134520cb","identifierType":"int","vectorType":"vector","partitions":4,"paramMap":{"numReplicas":0,"queryIdentifierCol":"id","featuresCol":"embedding","similarityThreshold":-1.0,"outputFormat":"full","predictionCol":"prediction","excludeSelf":false,"k":200}}

the file list

023-06-12 16:31:31     516928 devs/gnn/test/inference/metal/ranking/tfmodel/indices/3
2023-06-12 16:31:31          0 devs/gnn/test/inference/metal/ranking/tfmodel/metadata/_SUCCESS
2023-06-12 16:31:30        420 devs/gnn/test/inference/metal/ranking/tfmodel/metadata/part-00000

**Comment by jelmerk** [comment]
how many distinct values for id are there in the dataset ?

**Comment by jelmerk** [comment]
ok can you get check the executor logs for lines logged at info level that start with `partition 00`

**Comment by jelmerk** [comment]
yes

**Comment by perrital** [comment]
Yes they are there, each executor with its own partition and proportional number of items corresponding to the number of records spark UI indicates

**Comment by jelmerk** [comment]
I can't really think of any good reason for this happening unless some external force is removing files in the /tmp/tal folder

**Comment by perrital** [comment]
I'll keep investigating some more, to see it I can resolve this. Thanks for your help! really appreciate it :)

**Comment by jelmerk** [comment]
I can release a snapshot with some more logging if that helps. But I don't have a k8s cluster at my disposal to test on right now

**Comment by perrital** [comment]
found the problem. when saving and querying each executor gets a partition id which need not be the same as the partition id used for generating the index file on the executor (as the executors are assigned with the partitions randomly) 
I wonder, how it could work in the first place? what am I missing?

**Comment by jelmerk** [comment]
each index is saved as a file to the filessystem. eg `devs/gnn/test/inference/metal/ranking/tfmodel/indices/3` is the index for partition 3

On querying it loads the index from disk again.

This works if you use a distributed filesystem like hdfs or s3 but I guess you are not writing to a distributed filesystem but to local disk. And yeah that would not work

**Comment by perrital** [comment]
I don't understand, I actually use s3. I see that in creating the index the file is persisted on the executor machine. 
only when I do model.write().overwrite().save(path/in/s3)  it (tries) to persist it to s3

**Comment by jelmerk** [comment]
What does this return ?

`spark.sparkContext._jsc.hadoopConfiguration().get("fs.defaultFS")`

**Comment by jelmerk** [comment]
closing because it's a configuration issue
